File: src/lib/blockchain/adapters/tezos.js
/**
 * Tezos Adapter — CryptoLand
 * ===========================
 * Covers Tezos mainnet + Shadownet via TZIP-10 / Beacon (@airgap/beacon-sdk),
 * which every Tezos wallet speaks — Temple, Kukai, Umami, Naan, AirGap.
 * Connect + address + purchase signing work today with no contract deployed;
 * NFT minting (an FA2 / TZIP-12 contract call) is stubbed until VITE_CONTRACT_TEZOS
 * holds an originated KT1… address.
 *
 * Implements the universal BlockchainAdapter interface (same surface as evm.js).
 *
 * Five things make Tezos unlike the other families here:
 *   1. GHOSTNET IS GONE. Every ghostnet host (ghostnet.tzkt.io, rpc.tzkt.io/ghostnet,
 *      ghostnet.ecadinfra.com, ghostnet.teztnets.com) is dead in DNS or 404s, and it
 *      no longer appears in teztnets.json. Beacon's NetworkType enum still carries a
 *      GHOSTNET member — it is a trap. SHADOWNET is the current app-testing network.
 *   2. There is NO window.* injection for Beacon. It reaches extensions over
 *      postMessage and mobile wallets over a P2P matrix relay, so wallets cannot be
 *      feature-detected synchronously — Beacon's own pairing modal is the picker.
 *   3. A signable payload must be PACKED Micheline (a "05…" hex string). Wallets
 *      show a raw hex blob (or refuse) for anything else. See packMichelineString().
 *   4. Tezos has no ecrecover: a signature alone does not yield the signer's key,
 *      and an unrevealed account has no public key on-chain either. We therefore
 *      return the wallet's publicKey alongside every signature — the backend needs
 *      it for @taquito/utils verifySignature(payload, publicKey, signature).
 *   5. Minting REQUIRES an originated contract. Unlike Algorand ASAs or Stellar
 *      assets there is no native-asset shortcut: an NFT is an FA2 token with
 *      supply 1 plus TZIP-16/21 metadata in the token_metadata big_map.
 *
 * Native XTZ has 6 decimals (1 XTZ = 1,000,000 mutez); Beacon amounts are mutez
 * strings, never tez.
 *
 * The SDK is an optional peer dep and is lazy-loaded; the app builds and runs with
 * @airgap/beacon-sdk absent. (Note for whoever installs it: the Beacon bundle
 * expects Node globals — a Buffer/global polyfill may be needed in the Vite build.)
 */

import { ACTIVE_CHAIN } from '../config.js'
import { tileTokenId, tokenIdToTile, hasContract, mintStub } from './_shared.js'

export { tileTokenId, tokenIdToTile }

// ACTIVE_CHAIN.id is the Tezos chain_id (NetXdQprcVkpaWU / NetXsqzbfFenSTS), which
// is NOT what Beacon wants — it takes a NetworkType string. Derive it from the
// build's testnet flag rather than mapping chain ids, and use the literal rather
// than NetworkType.SHADOWNET: SDK builds older than 4.x lack that enum member and
// would silently hand the wallet `undefined`. The enum values ARE these strings.
const NETWORK_TYPE = ACTIVE_CHAIN.testnet ? 'shadownet' : 'mainnet'
const NETWORK      = { type: NETWORK_TYPE, rpcUrl: ACTIVE_CHAIN.rpcUrl }
const RPC          = ACTIVE_CHAIN.rpcUrl.replace(/\/+$/, '')

// Validation-pass index of manager operations (transactions, originations).
// 0 = consensus, 1 = voting, 2 = anonymous, 3 = manager.
const MANAGER_PASS = 3
// Hard cap on the block catch-up scan in waitForTx so a stalled tab can never fan
// out into hundreds of RPC fetches when it resumes.
const MAX_SCAN = 12

let _sdk       = null   // cached @airgap/beacon-sdk module namespace
let _client    = null   // the single DAppClient instance
let _address   = null
let _publicKey = null

const _cbs      = { accounts: [], chain: [], disconnect: [] }
let   _subscribed = false
let   _lastNetwork = NETWORK_TYPE

// ── SDK loading (optional peer dep) ──────────────────────────────────────────

async function beacon() {
  if (_sdk) return _sdk
  if (typeof window === 'undefined') {
    throw new Error('Tezos wallet access is browser-only (Beacon needs postMessage / a P2P relay).')
  }
  let mod
  try {
    mod = await import('@airgap/beacon-sdk')
  } catch {
    throw new Error(
      'Beacon SDK not available — install it with: npm i @airgap/beacon-sdk@^4.8.1 ' +
      '(Tezos wallets inject no window global, so the package is required).'
    )
  }
  _sdk = mod?.DAppClient ? mod : (mod?.default ?? mod)
  if (!_sdk?.DAppClient) throw new Error('Beacon SDK loaded but exposes no DAppClient — check the installed version.')
  return _sdk
}

/**
 * One DAppClient per page. A second instance fights the first over the same
 * localStorage/IndexedDB keys and can drop the active account mid-session; the
 * constructor also touches storage, so it must never run at module load or in SSR.
 */
async function client() {
  if (_client) return _client
  const { DAppClient } = await beacon()
  _client = new DAppClient({ name: 'CryptoLand', network: NETWORK, preferredNetwork: NETWORK_TYPE })
  return _client
}

// ── Address / hash validation ───────────────────────────────────────────────

/**
 * base58check, 36 chars. Implicit: tz1 (ed25519), tz2 (secp256k1), tz3 (P-256),
 * tz4 (BLS). Originated contracts are KT1…. Base58 has no 0, O, I or l.
 */
function assertAddress(addr) {
  if (typeof addr !== 'string' || !addr) throw new Error('Tezos wallet returned no address.')
  if (/^(edsk|spsk|p2sk|edesk|BLsk)/.test(addr)) {
    throw new Error('Refusing a Tezos SECRET key — CryptoLand only ever handles public addresses.')
  }
  if (!/^(tz[1-4]|KT1)[1-9A-HJ-NP-Za-km-z]{33}$/.test(addr)) {
    throw new Error(`Not a valid Tezos address: ${addr}`)
  }
  return addr
}

// ── Micheline payload packing ───────────────────────────────────────────────

/**
 * Build the standard TZIP-10 signed-payload envelope for a UTF-8 string:
 *
 *   05        packed-data prefix (what `PACK` emits; wallets only render "05…" safely)
 *   01        Micheline primitive tag for `string`
 *   xxxxxxxx  4-byte BIG-ENDIAN length of the payload — in BYTES, not characters
 *   …         hex of the UTF-8 bytes
 *
 * The byte-length distinction is the live trap: our purchase text contains an
 * em dash, which is 3 UTF-8 bytes, so a length taken from `text.length` produces a
 * payload the wallet renders as garbage and the backend cannot verify.
 * Verify server-side with @taquito/utils verifySignature(payload, publicKey, signature).
 */
function packMichelineString(text) {
  const bytes = new TextEncoder().encode(text)
  let hex = ''
  for (const b of bytes) hex += b.toString(16).padStart(2, '0')
  return `0501${bytes.length.toString(16).padStart(8, '0')}${hex}`
}

// ── Wallet discovery ─────────────────────────────────────────────────────────

export function detectWallets() {
  // Beacon wallets are NOT feature-detectable: extensions answer an async
  // postMessage handshake and mobile wallets pair over a relay, so nothing lands
  // on window. Beacon's pairing modal enumerates what is actually installed —
  // this list only seeds the UI. (Temple's legacy window.temple object belongs to
  // its own pre-TZIP-10 API and says nothing about Beacon support.)
  return [
    { id: 'temple', name: 'Temple',  icon: '🏛️' },
    { id: 'kukai',  name: 'Kukai',   icon: '🌐' },
    { id: 'umami',  name: 'Umami',   icon: '🍜' },
    { id: 'naan',   name: 'Naan',    icon: '🫓' },
    { id: 'airgap', name: 'AirGap',  icon: '✈️' },
  ]
}

// ── Wallet connection ───────────────────────────────────────────────────────

/**
 * Beacon PERSISTS the active account in localStorage, so after a reload the app is
 * still paired. Reuse that account instead of calling requestPermissions() again —
 * a second call re-opens the pairing modal for an already-connected user.
 * Note requestPermissions() must run inside a user gesture or the popup is blocked.
 */
export async function connect() {
  const c = await client()

  let account = await c.getActiveAccount().catch(() => null)
  if (!account?.address) {
    const perms = await c.requestPermissions({ network: NETWORK })
    account = (await c.getActiveAccount().catch(() => null)) ?? perms
  }

  const address = account?.address ?? account?.accountInfo?.address
  if (!address) throw new Error('Tezos wallet connection rejected')
  _address   = assertAddress(address)
  _publicKey = account?.publicKey ?? account?.accountInfo?.publicKey ?? null

  assertNetwork(account)
  _lastNetwork = account?.network?.type ?? NETWORK_TYPE
  return { address: _address, chainId: ACTIVE_CHAIN.id, chainName: ACTIVE_CHAIN.name }
}

/**
 * A wallet left on the wrong network signs happily and only fails at injection,
 * after the user has approved. 'custom' is tolerated: that is what wallets report
 * when the user points them at a self-hosted RPC for this same network.
 */
function assertNetwork(account) {
  const type = account?.network?.type
  if (!type || type === 'custom' || type === NETWORK_TYPE) return
  throw new Error(
    `Wallet is on Tezos "${type}" but this build targets ${ACTIVE_CHAIN.name} (${NETWORK_TYPE}). ` +
    'Switch networks in the wallet and reconnect.'
  )
}

export function disconnect() {
  // BOTH calls are required. disconnect() only tears down the transport (P2P /
  // postMessage); the account stays in localStorage and the app comes back
  // "connected" on the next reload. clearActiveAccount() is what actually forgets it.
  _client?.disconnect?.().catch(() => {})
  _client?.clearActiveAccount?.().catch(() => {})
  _address   = null
  _publicKey = null
}

export function getAddress() {
  return _address
}

export function getChainId() {
  return ACTIVE_CHAIN.id   // Tezos chain_id, e.g. NetXdQprcVkpaWU — not a number
}

export async function switchChain() { /* a Tezos build targets one network */ }

// ── Signing ─────────────────────────────────────────────────────────────────

async function signText(text) {
  const c = await client()
  const account = await c.getActiveAccount().catch(() => null)
  const source = account?.address ?? _address
  if (!source) throw new Error('Connect a Tezos wallet first.')
  _publicKey = account?.publicKey ?? _publicKey

  const { SigningType } = await beacon()
  const payload = packMichelineString(text)
  const res = await c.requestSignPayload({
    // MICHELINE is the only type wallets display as readable text. SigningType.RAW
    // is shown as an opaque blob and several wallets refuse it outright, and
    // OPERATION is reserved for forged operation bytes — never reuse it for login.
    signingType:   SigningType?.MICHELINE ?? 'micheline',
    payload,
    sourceAddress: source,
  })
  const signature = res?.signature
  if (!signature) throw new Error('Tezos wallet returned no signature.')

  // publicKey can legitimately be missing when the account has never been revealed
  // on-chain and the wallet withheld it at permission time — say so rather than
  // letting the backend fail an unexplained verifySignature().
  return {
    signature,
    address:   source,
    publicKey: _publicKey,
    payload,
    message:   text,
  }
}

export async function signMessage(message) {
  const { signature, address, publicKey, payload } = await signText(message)
  return { signature, address, publicKey, payload }
}

// ── Purchase signature (proof of wallet control, no contract needed) ─────────

export async function signPurchase({ tileKey, price }) {
  // requestSignPayload binds nothing about the network into the payload, so a
  // Shadownet signature would otherwise verify against a mainnet purchase. Write
  // the chain_id into the text ourselves.
  const text =
    `CryptoLand purchase — tile ${tileKey} for $${price} — ${_address ?? ''}\n` +
    `Chain: ${ACTIVE_CHAIN.id}`
  const { signature, publicKey, payload, address } = await signText(text)
  return { signature, message: text, payload, publicKey, address }
}

// ── Native payment (a plain XTZ transfer — no contract involved) ─────────────

/**
 * Pay for a tile in XTZ, from the user's own wallet.
 *
 * A bare transaction operation to the treasury, NOT a call into the FA2
 * contract: minting is gated on an originated KT1 that most builds do not have,
 * and the price is per-tile anyway, which no fixed on-chain price can express.
 * This path therefore works with VITE_CONTRACT_TEZOS unset.
 *
 * `amount` is a decimal STRING of MUTEZ (1 XTZ = 1,000,000 mutez) straight from
 * the server's quote. It stays a string/BigInt end to end — Beacon wants a
 * string here, and a Number would round large quotes rather than fail loudly.
 */
export async function payNative({ to, amount, from }) {
  if (!to)     throw new Error('No treasury address for this chain')
  if (!amount) throw new Error('No amount to pay')

  const mutez = BigInt(amount)          // throws on a malformed quote
  if (mutez <= 0n) throw new Error('Refusing to send a non-positive amount')

  // tz1-4 (implicit) and KT1 (originated) are both legitimate treasuries — a KT1
  // just runs its default entrypoint on receipt.
  assertAddress(to)

  const c = await client()
  const account = await c.getActiveAccount().catch(() => null)
  const payer = account?.address ?? from ?? _address
  if (!payer) throw new Error('Connect a Tezos wallet before paying.')

  // Beacon signs with the wallet's ACTIVE account and an operation request
  // carries no per-operation source override. If the user switched accounts
  // since connecting, the payment would leave an address the backend never
  // associated with this purchase — say so rather than paying from the wrong tz1.
  if (from && from !== payer) {
    throw new Error(
      `The Tezos wallet's active account is ${payer}, not ${from}. ` +
      'Switch back in the wallet (or reconnect) and retry the payment.'
    )
  }

  // A wallet left on the wrong network signs happily and only fails at injection,
  // after the user has approved — and on mainnet that is a real payment.
  assertNetwork(account)

  const out = await c.requestOperation({
    operationDetails: [{
      kind:        'transaction',
      destination: to,
      // MUTEZ, as a string — Beacon takes no numbers here. fee/gas_limit/
      // storage_limit are deliberately omitted: the wallet simulates the
      // operation and fills them, and hand-set limits are the usual cause of a
      // "gas exhausted" rejection.
      amount:      mutez.toString(),
    }],
  })

  const txHash = out?.transactionHash ?? out?.opHash ?? null
  if (!txHash) throw new Error('Tezos wallet returned no operation hash for the payment.')
  return { txHash, from: payer }
}

/** Whether this build can take a wallet payment at all. */
export function supportsNativePay() {
  // Beacon injects nothing into window — as detectWallets() notes, extensions
  // answer an async postMessage handshake and mobile wallets pair over a relay,
  // so there is no wallet to feature-detect synchronously. The one real gate is
  // the runtime: DAppClient needs postMessage / a P2P relay / localStorage, none
  // of which exist under SSR or in the test runner.
  if (ACTIVE_CHAIN.gasless || ACTIVE_CHAIN.halted) return false
  return typeof window !== 'undefined'
}

// ── NFT mint (stubbed until an FA2 contract is originated) ───────────────────

export async function mintTile({ tx, ty, country, toAddress }) {
  if (!hasContract()) return mintStub('Tezos FA2 contract not originated')

  const c = await client()
  const account = await c.getActiveAccount().catch(() => null)
  if (!account?.address) throw new Error('Connect a Tezos wallet to mint.')
  assertNetwork(account)

  const tokenId = tileTokenId(tx, ty)   // Michelson `nat` is unbounded — no packing needed

  // The backend emits the Micheline JSON for the contract's mint entrypoint. It
  // cannot be built generically: FA2 (TZIP-12) standardises transfer/balance_of/
  // update_operators but NOT mint, so the argument shape is contract-specific.
  const BASE = import.meta.env.VITE_API_BASE ?? ''
  const res  = await fetch(`${BASE}/tezos/build-mint`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify({
      tx, ty, country,
      owner:    toAddress,
      source:   account.address,
      tokenId:  String(tokenId),
      contract: ACTIVE_CHAIN.contractAddress,
      network:  NETWORK_TYPE,
    }),
  })
  if (!res.ok) throw new Error('Failed to build Tezos mint operation')
  const { parameters } = await res.json()
  if (!parameters?.entrypoint) throw new Error('Backend returned no Micheline parameters for the FA2 mint.')

  const out = await c.requestOperation({
    operationDetails: [{
      kind:        'transaction',
      destination: ACTIVE_CHAIN.contractAddress,
      // MUTEZ, as a string. '0' because minting transfers no XTZ — the wallet
      // fills fee/gas_limit/storage_limit itself after simulating, and hand-set
      // limits are the usual cause of a "gas exhausted" rejection.
      amount:      '0',
      parameters,
    }],
  })

  const txHash = out?.transactionHash ?? out?.opHash ?? null
  if (!txHash) throw new Error('Tezos wallet returned no operation hash.')
  return { txHash, tokenId: String(tokenId), minted: true }
}

// ── Marketplace (activates with the originated contract) ─────────────────────

export async function listForSale() { throw new Error('Tezos marketplace: available after FA2 contract deploy') }
export async function unlistTile()   { throw new Error('Tezos marketplace: available after FA2 contract deploy') }
export async function buyTile()      { throw new Error('Tezos marketplace: available after FA2 contract deploy') }

// ── Reads (delegate to backend / indexer) ───────────────────────────────────

export async function ownerOf()          { return null }
export async function getTileData()      { return null }
export async function getOwnedTokenIds() { return [] }
export async function totalSupply()      { return 0 }

// ── Confirmation ────────────────────────────────────────────────────────────

async function rpcJson(path) {
  const r = await fetch(`${RPC}${path}`, { headers: { Accept: 'application/json' } }).catch(() => null)
  if (!r?.ok) return null
  return r.json().catch(() => null)
}

/**
 * An operation group is included atomically but each content carries its OWN
 * result, and a `mint` that calls another contract fails inside
 * internal_operation_results — a group can sit in a block having done nothing.
 * Anything other than 'applied' (failed / backtracked / skipped) is a failure.
 */
function opResult(group) {
  for (const content of group?.contents ?? []) {
    const result = content?.metadata?.operation_result
    if (result?.status && result.status !== 'applied') return { ok: false, status: result.status, errors: result.errors }
    for (const internal of content?.metadata?.internal_operation_results ?? []) {
      const r = internal?.result
      if (r?.status && r.status !== 'applied') return { ok: false, status: r.status, errors: r.errors }
    }
  }
  return { ok: true }
}

/**
 * Polls the node directly rather than an indexer: TzKT has no verified API host
 * for Shadownet (api.ghostnet.tzkt.io is gone with the network), while the RPC in
 * ACTIVE_CHAIN.rpcUrl is guaranteed to exist for whichever network this build targets.
 * Resolves on inclusion + 'applied'; Tenderbake finality is 2 further blocks (~16s)
 * and is left to the backend's confirmation policy.
 */
export async function waitForTx(opHash, maxWait = 90_000) {
  if (!opHash) throw new Error('waitForTx: no Tezos operation hash given.')
  const start = Date.now()
  let next = null

  while (Date.now() - start < maxWait) {
    const header = await rpcJson('/chains/main/blocks/head/header')
    const head = typeof header?.level === 'number' ? header.level : null

    if (head !== null) {
      // Start two blocks back — the op may already have landed between injection
      // and the first poll — then walk forward, never re-fetching a block.
      if (next === null) next = Math.max(head - 2, 1)
      if (head - next > MAX_SCAN) next = head - MAX_SCAN

      for (; next <= head; next++) {
        const groups = await rpcJson(`/chains/main/blocks/${next}/operations/${MANAGER_PASS}`)
        if (!Array.isArray(groups)) break   // RPC hiccup — retry this same level next poll
        const group = groups.find(g => g?.hash === opHash)
        if (!group) continue
        const { ok, status, errors } = opResult(group)
        if (!ok) throw new Error(`Tezos operation ${opHash} ${status}: ${errors?.[0]?.id ?? 'no details'}`)
        return group
      }
    }
    await new Promise(r => setTimeout(r, 3000))
  }
  throw new Error(`Tezos operation ${opHash} not confirmed within ${maxWait / 1000}s`)
}

// ── Listeners ───────────────────────────────────────────────────────────────

/**
 * Beacon exposes exactly one useful account signal, ACTIVE_ACCOUNT_SET, and fires
 * it with `undefined` when the account is cleared — that is also the only
 * disconnect notice we get. One subscription feeds all three listener kinds
 * because subscribeToEvent() returns no unsubscribe handle in v4: removeListeners()
 * can only drop our own callbacks, so the handler must stay tolerant of an empty list.
 */
async function subscribe() {
  if (_subscribed || typeof window === 'undefined') return
  const c = await client().catch(() => null)
  const { BeaconEvent } = (await beacon().catch(() => ({}))) ?? {}
  const event = BeaconEvent?.ACTIVE_ACCOUNT_SET ?? 'ACTIVE_ACCOUNT_SET'
  if (typeof c?.subscribeToEvent !== 'function') return   // older SDK: callbacks stay idle rather than crash
  _subscribed = true
  try {
    await c.subscribeToEvent(event, (account) => {
      const address = account?.address ?? null
      _address   = address
      _publicKey = account?.publicKey ?? null
      _cbs.accounts.forEach(cb => { try { cb(address) } catch { /* listener owns its errors */ } })
      if (!address) _cbs.disconnect.forEach(cb => { try { cb() } catch { /* ignore */ } })

      const network = account?.network?.type
      if (network && network !== _lastNetwork) {
        _lastNetwork = network
        _cbs.chain.forEach(cb => { try { cb(network) } catch { /* ignore */ } })
      }
    })
  } catch {
    _subscribed = false
  }
}

export async function onAccountsChanged(cb) {
  if (typeof cb !== 'function') return
  _cbs.accounts.push(cb)
  await subscribe()
}

// Not a no-op despite the single-network build: a user who flips the wallet to
// another network would sign operations this deployment can never inject.
export async function onChainChanged(cb) {
  if (typeof cb !== 'function') return
  _cbs.chain.push(cb)
  await subscribe()
}

export async function onDisconnect(cb) {
  if (typeof cb !== 'function') return
  _cbs.disconnect.push(cb)
  await subscribe()
}

export function removeListeners() {
  _cbs.accounts.length   = 0
  _cbs.chain.length      = 0
  _cbs.disconnect.length = 0
}

export const ADAPTER_TYPE = 'tezos'

File: server/verifiers/tezos.py
"""
verifiers/tezos.py — confirm an XTZ payment landed in the treasury.

── The endpoint problem, stated plainly ──────────────────────────────────────
`chain.rpcs` for Tezos is [rpc.tzkt.io/mainnet, prod.tcinfra.net/rpc/mainnet].
Both are **bare Tezos node RPCs**, and a bare node RPC has no "fetch operation
by hash" route at all. An operation lives inside a block, so looking one up
means already knowing its level — the node is content-addressed by block, not by
operation. That is why adapters/tezos.js `waitForTx()` scans blocks rather than
querying a hash.

The **TzKT API** (`https://api.tzkt.io/v1/operations/transactions/{hash}`) is
built for exactly this question, and is run by the same operator as the
`rpc.tzkt.io` host already trusted in `config.js`.

🔴 **`api.tzkt.io` is NOT in `config.js` and is therefore NOT covered by
`scripts/check-rpcs.mjs`.** It is a real, undeclared third-party dependency of
the Tezos payment rail. Two things reduce the exposure but do not remove it:

  1. a transport failure here is `pending`, never a rejection, so a TzKT outage
     delays settlement instead of telling a buyer they did not pay; and
  2. `_verify_via_node()` below falls back to the CONFIGURED node RPCs with the
     same bounded block scan the adapter uses, so a recent payment still
     verifies with TzKT completely unavailable.

The fallback is bounded (`_MAX_SCAN` blocks) so a stalled poller can never fan
out into hundreds of RPC fetches. It covers the case that matters — a buyer who
paid moments ago — and returns `pending` for anything older, which is the
correct answer while our only hash index is down.

── Why the whole operation GROUP is summed ───────────────────────────────────
One Tezos operation hash covers a group: the manager operation the wallet signed
plus every `internal_operation_result` it triggered. A single hash can credit
several addresses — verified live against
`opTCGgQL7dK7dQHGWRrG75zPNieyZ8W1c9jdMZPh7WvipCTNrpo`, which pays 5000000 mutez
to a KT1 and then internally forwards 250000, 500000 and 4250000 mutez to three
different tz-addresses. Only the contents addressed to the treasury are counted.
"""
from __future__ import annotations

import re
from urllib.parse import urlparse

import aiohttp

from chain_pay import HEADERS, HTTP_TIMEOUT, PaymentProof, from_base_units
from chain_registry import Chain

# Validation-pass index of manager operations (transactions, originations) in a
# block. 0 = consensus, 1 = voting, 2 = anonymous, 3 = manager. Same constant
# adapters/tezos.js uses.
_MANAGER_PASS = 3

# How far back the node-RPC fallback will walk. Tezos blocks are ~8s, so 12
# blocks is ~1.6 minutes — comfortably longer than the gap between a wallet
# injecting a payment and the first verify poll.
_MAX_SCAN = 12

# base58check, 51 chars, "o" prefix. Base58 excludes 0, O, I and l.
_OPHASH_RE = re.compile(r"^o[1-9A-HJ-NP-Za-km-z]{50}$")


def _tzkt_api(chain: Chain) -> str:
    """
    The TzKT API host matching this build's network, derived from the configured
    node RPC rather than hard-coded, so a testnet build does not silently query
    mainnet. Mainnet (`https://api.tzkt.io/v1`) is the only one verified live;
    the `api.<net>.tzkt.io` shape for other networks follows TzKT's documented
    convention but is unverified here, and Shadownet in particular has had no
    working TzKT host since Ghostnet was retired (see adapters/tezos.js).
    """
    net = ""
    for raw in chain.rpcs:
        parts = [p for p in urlparse(raw or "").path.split("/") if p]
        if parts:
            net = parts[-1].lower()
            break
    if not net or net == "mainnet":
        return "https://api.tzkt.io/v1"
    return f"https://api.{net}.tzkt.io/v1"


async def _get_json(session: aiohttp.ClientSession, url: str):
    """GET returning (status, parsed-json-or-None). Raises only on transport."""
    async with session.get(url, headers=HEADERS, timeout=HTTP_TIMEOUT) as r:
        if r.status in (400, 404):
            # TzKT answers 400 for a hash that is not base58check-shaped and 404
            # for an unknown route; neither is a node failure.
            return r.status, None
        if r.status != 200:
            raise RuntimeError(f"HTTP {r.status}")
        return 200, await r.json(content_type=None)


def _settle(chain: Chain, paid: int, payer: str, confs: int, min_amount: int) -> PaymentProof:
    """The amount/confirmation verdict, shared by both lookup paths."""
    if paid < min_amount:
        return PaymentProof(
            ok=False,
            paid=paid,
            payer=payer,
            confirmations=confs,
            error=(
                f"Underpaid: treasury received {from_base_units(paid, chain.decimals):.6f} "
                f"{chain.symbol}, expected at least "
                f"{from_base_units(min_amount, chain.decimals):.6f}"
            ),
        )
    if confs < chain.confirmations:
        return PaymentProof(
            ok=False, pending=True, paid=paid, payer=payer, confirmations=confs,
            error=f"Confirming ({confs}/{chain.confirmations})",
        )
    return PaymentProof(ok=True, paid=paid, payer=payer, confirmations=confs)


# ── Path 1: the TzKT indexer ────────────────────────────────────────────────

async def _verify_via_tzkt(
    chain: Chain, op_hash: str, treasury: str, min_amount: int,
    session: aiohttp.ClientSession,
) -> PaymentProof | None:
    """
    Verdict via TzKT, or None to mean "TzKT could not answer — try the node".

    None is deliberately distinct from a pending PaymentProof: it hands control
    to the fallback, whereas a pending proof is already the final answer.
    """
    api = _tzkt_api(chain)

    status, ops = await _get_json(session, f"{api}/operations/transactions/{op_hash}")
    if status != 200 or ops is None:
        return None
    if not isinstance(ops, list) or not ops:
        # A well-formed hash TzKT has never seen returns []. Not indexed yet.
        return None

    paid = 0
    payer = ""
    level = 0
    saw_failed_to_treasury = False
    for op in ops:
        # Tezos addresses are base58check and MIXED CASE — tz1/tz2/tz3/tz4 and
        # KT1 differ from each other only in that prefix, and the trailing bytes
        # are a checksum over the raw key hash. `.lower()` would turn a valid
        # address into a string that decodes to nothing, so every comparison
        # would fail and no payment would ever verify. Compare as given.
        if (op.get("target") or {}).get("address") != treasury:
            continue
        if op.get("status") != "applied":
            # Tezos backtracks an entire group when any content fails, so this
            # means the transfer did not happen — a permanent verdict, not a wait.
            saw_failed_to_treasury = True
            continue
        paid += int(op.get("amount") or 0)
        level = level or int(op.get("level") or 0)
        if not payer:
            # `initiator` is the tz1 that signed the group; it is present only on
            # INTERNAL operations, where `sender` is the forwarding contract.
            # Prefer it so `payer` is always the human who paid.
            payer = (
                (op.get("initiator") or {}).get("address")
                or (op.get("sender") or {}).get("address")
                or ""
            )

    if paid <= 0:
        if saw_failed_to_treasury:
            return PaymentProof(
                ok=False,
                error="The transfer to the treasury did not succeed on-chain",
            )
        return PaymentProof(ok=False, error="Payment was not sent to the treasury address")

    status, head = await _get_json(session, f"{api}/head")
    if status != 200 or head is None:
        return None
    # TzKT reports which chain it is indexing. Free, and it is what stops a
    # testnet index from settling a mainnet sale.
    chain_id = head.get("chainId")
    if chain_id and chain.chain_id and chain_id != chain.chain_id:
        return None
    confs = max(0, int(head.get("level") or 0) - level + 1) if level else 0

    return _settle(chain, paid, payer, confs, min_amount)


# ── Path 2: the configured node RPCs, bounded block scan ────────────────────

def _sum_group(group: dict, treasury: str) -> tuple[int, str, bool]:
    """
    (mutez credited to the treasury, payer, saw-a-failed-transfer) for one
    operation group as a bare node RPC returns it.

    Node-RPC amounts are STRINGS; internal transfers live under
    `metadata.internal_operation_results` with their status under `result`,
    not `operation_result`.
    """
    paid = 0
    payer = ""
    failed = False
    for content in group.get("contents") or []:
        meta = content.get("metadata") or {}
        if content.get("kind") == "transaction" and content.get("destination") == treasury:
            if (meta.get("operation_result") or {}).get("status") == "applied":
                paid += int(content.get("amount") or 0)
                payer = payer or content.get("source") or ""
            else:
                failed = True
        for internal in meta.get("internal_operation_results") or []:
            if internal.get("kind") != "transaction" or internal.get("destination") != treasury:
                continue
            if (internal.get("result") or {}).get("status") == "applied":
                paid += int(internal.get("amount") or 0)
                # The group's own signer, not the forwarding contract in
                # `internal["source"]`.
                payer = payer or (content.get("source") or "")
            else:
                failed = True
    return paid, payer, failed


async def _verify_via_node(
    chain: Chain, op_hash: str, treasury: str, min_amount: int,
    session: aiohttp.ClientSession,
) -> PaymentProof:
    """Fallback for when TzKT is unreachable — see the module docstring."""
    last_err = ""
    for raw in chain.rpcs:
        base = (raw or "").rstrip("/")
        if not base:
            continue
        try:
            status, header = await _get_json(session, f"{base}/chains/main/blocks/head/header")
            if status != 200 or header is None:
                last_err = "Node returned no head header"
                continue
            if header.get("chain_id") and chain.chain_id and header["chain_id"] != chain.chain_id:
                last_err = f"Node at {base} is on '{header['chain_id']}', not {chain.chain_id}"
                continue
            head = int(header.get("level") or 0)
            if not head:
                last_err = "Node head has no level"
                continue

            for level in range(head, max(head - _MAX_SCAN, 0), -1):
                status, groups = await _get_json(
                    session, f"{base}/chains/main/blocks/{level}/operations/{_MANAGER_PASS}"
                )
                if status != 200 or not isinstance(groups, list):
                    continue
                group = next((g for g in groups if g.get("hash") == op_hash), None)
                if group is None:
                    continue
                paid, payer, failed = _sum_group(group, treasury)
                if paid <= 0:
                    if failed:
                        return PaymentProof(
                            ok=False,
                            error="The transfer to the treasury did not succeed on-chain",
                        )
                    return PaymentProof(ok=False, error="Payment was not sent to the treasury address")
                return _settle(chain, paid, payer, max(0, head - level + 1), min_amount)

            last_err = f"Operation not in the last {_MAX_SCAN} blocks"
        except Exception as exc:
            last_err = str(exc)
            continue

    return PaymentProof(ok=False, pending=True, error=last_err or "No Tezos endpoint responded")


# ── The contract ────────────────────────────────────────────────────────────

async def verify(
    chain: Chain,
    tx_hash: str,
    treasury: str,
    min_amount: int,
    session: aiohttp.ClientSession,
) -> PaymentProof:
    """
    Confirm an XTZ transfer credited the treasury.

    `tx_hash` is what `payNative()` in adapters/tezos.js returns: the Beacon
    OPERATION hash (`out.transactionHash ?? out.opHash`) — base58check with an
    "o" prefix, 51 characters. It is not a hex hash and must not be case-folded.

    The amount is `amount` in MUTEZ (1 XTZ = 1,000,000), read straight off the
    operation. Unlike Stellar there is no decimal-string conversion to get wrong:
    both TzKT and the node RPC report integers (the node reports them as
    integer-valued strings).

    Only `status == "applied"` counts. Anything else — failed, backtracked,
    skipped — means the group was rolled back and no XTZ moved.
    """
    op_hash = (tx_hash or "").strip()
    treasury = (treasury or "").strip()
    if not _OPHASH_RE.match(op_hash):
        # Pending rather than rejected: a malformed hash is much more likely a
        # value truncated in flight than a false claim we should deny outright.
        return PaymentProof(
            ok=False, pending=True,
            error="Not a Tezos operation hash yet (expected 51 base58 characters starting with 'o')",
        )
    if not treasury:
        return PaymentProof(ok=False, error="No treasury address configured")

    try:
        verdict = await _verify_via_tzkt(chain, op_hash, treasury, min_amount, session)
    except Exception:
        # A TzKT transport failure is not a verdict — fall through to the node.
        verdict = None
    if verdict is not None:
        return verdict

    return await _verify_via_node(chain, op_hash, treasury, min_amount, session)
